packages/s3-store/test.ts

```diff
@@ -113,6 +113,41 @@ describe('S3DataStore', function () {
     }
   })
 
+  it('completes an incomplete part when deferred length becomes resolved', async function () {
+    const store = this.datastore
+    const incompleteSize = 2 * 1024 * 1024 // 2MB
+    const uploadIncompletePart = sinon.spy(store, 'uploadIncompletePart')
+    const uploadPart = sinon.spy(store, 'uploadPart')
+    const finishMultipartUpload = sinon.spy(store, 'finishMultipartUpload')
+    const upload = new Upload({
+      id: 'deferred-incomplete-part-test-' + Uid.rand(),
+      // Deferred length
+      size: undefined,
+      offset: 0,
+    })
+    let offset = upload.offset
+    await store.create(upload)
+
+    // Upload a single chunk small enough to create an incomplete part
+    offset = await store.write(
+      Readable.from(Buffer.alloc(incompleteSize)),
+      upload.id,
+      offset
+    )
+    assert.equal(uploadIncompletePart.called, true)
+    assert.equal(uploadPart.called, false)
+
+    // Simulate the completion PATCH of a deferred length multipart upload
+    // Resolve the deferred length
+    await store.declareUploadLength(upload.id, incompleteSize)
+    // Notify the store to complete the multipart upload (empty payload)
+    await store.write(Readable.from(Buffer.alloc(0)), upload.id, offset)
+    // The incomplete part will now be completed
+    assert.equal(uploadPart.called, true)
+    // The multipart upload will now be completed
+    assert.equal(finishMultipartUpload.called, true)
+  })
+
```

Comment:
I prepared another test where we test if it works when there is a one complete part and one incomplete part.

```
  it('completes an incomplete part when deferred length becomes resolved', async function () {
    const store = this.datastore
    const firstPartSize = 8 * 1024 * 1024 // 8MB
    const secondPartSize = 2 * 1024 * 1024 // 2MB

    const uploadPart = sinon.spy(store, 'uploadPart')
    const getIncompletePart = sinon.spy(store, 'getIncompletePart')
    const uploadIncompletePart = sinon.spy(store, 'uploadIncompletePart')
    const finishMultipartUpload = sinon.spy(store, 'finishMultipartUpload')

    sinon.stub(store, 'calcOptimalPartSize').callsFake(() => firstPartSize)

    const upload = new Upload({
      id: 'deferred-incomplete-part-test-' + Uid.rand(),
      // Deferred length
      size: undefined,
      offset: 0,
    })

    await store.create(upload)

    // Upload a single chunk large enough to create a part
    {
      upload.offset = await store.write(
        Readable.from(Buffer.alloc(firstPartSize)),
        upload.id,
        upload.offset
      )
      assert.equal(uploadPart.called, true)
      assert.equal(uploadIncompletePart.called, false)
    }

    // Upload a single chunk small enough to create an incomplete part
    {
      uploadPart.resetHistory()
      uploadIncompletePart.resetHistory()

      upload.offset = await store.write(
        Readable.from(Buffer.alloc(secondPartSize)),
        upload.id,
        upload.offset
      )
      assert.equal(uploadIncompletePart.called, true)
      assert.equal(uploadPart.called, false)
    }

    // Simulate the completion PATCH of a deferred length multipart upload
    {
      getIncompletePart.resetHistory()
      uploadIncompletePart.resetHistory()

      // Resolve the deferred length
      await store.declareUploadLength(upload.id, firstPartSize + secondPartSize)
      // Notify the store to complete the multipart upload (empty payload)
      await store.write(Readable.from(Buffer.alloc(0)), upload.id, upload.offset)
      // Final write needs to check for incomplete part
      assert.equal(getIncompletePart.called, true)
      // The incomplete part will now be completed
      assert.equal(uploadPart.called, true)
      // The multipart upload will now be completed
      assert.equal(finishMultipartUpload.called, true)
    }
  })
```